src/apint/arithmetic.rs

```diff
@@ -53,154 +77,184 @@ impl ApInt {
         self.wrapping_inc();
         //`wrapping_inc` handles clearing the unused bits
     }
 
-    /// Adds `rhs` to `self` and returns the result.
-    /// 
-    /// **Note:** This will **not** allocate memory.
-    /// 
-    /// # Errors
-    /// 
-    /// - If `self` and `rhs` have unmatching bit widths.
-    pub fn into_wrapping_add(self, rhs: &ApInt) -> Result<ApInt> {
-        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_add_assign)
+    /// Negates this `ApInt` and returns the result.
+    pub fn into_wrapping_neg(self) -> ApInt {
+        forward_mut_impl(self, ApInt::wrapping_neg)
     }
 
     /// Add-assigns `rhs` to `self` inplace.
     /// 
-    /// **Note:** This will **not** allocate memory.
-    /// 
     /// # Errors
     /// 
     /// - If `self` and `rhs` have unmatching bit widths.
     pub fn wrapping_add_assign(&mut self, rhs: &ApInt) -> Result<()> {
-        match self.zip_access_data_mut(rhs)? {
+        match self.zip_access_data_mut_self(rhs)? {
             Inl(lhs, rhs) => {
-                let lval = lhs.repr();
-                let rval = rhs.repr();
-                let result = lval.wrapping_add(rval);
-                *lhs = Digit(result);
+                *lhs = lhs.wrapping_add(rhs);
             }
             Ext(lhs, rhs) => {
-                let mut carry = Digit::zero();
-                for (l, r) in lhs.into_iter().zip(rhs) {
-                    *l = ll::carry_add(*l, *r, &mut carry);
+                let (temp, mut carry) = lhs[0].carrying_add(rhs[0]);
+                lhs[0] = temp;
+                for i in 1..lhs.len() {
+                    let temp = lhs[i].dd().wrapping_add(rhs[i].dd()).wrapping_add(carry.dd());
+                    lhs[i] = temp.lo();
+                    carry = temp.hi();
                 }
             }
         }
         self.clear_unused_bits();
-        // Maybe we should return a recoverable error upon carry != 0 at this point.
         Ok(())
     }
 
-    /// Subtracts `rhs` from `self` and returns the result.
+    /// Adds `rhs` to `self` and returns the result.
     /// 
-    /// # Note
+    /// # Errors
     /// 
-    /// In the low-level bit-wise representation there is no difference between signed
-    /// and unsigned subtraction of fixed bit-width integers. (Cite: LLVM)
+    /// - If `self` and `rhs` have unmatching bit widths.
+    pub fn into_wrapping_add(self, rhs: &ApInt) -> Result<ApInt> {
+        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_add_assign)
+    }
+
+    /// Add-assigns `rhs` to `self` inplace, and returns a boolean indicating if overflow occured,
+    /// according to the **unsigned** interpretation of overflow.
     /// 
     /// # Errors
     /// 
     /// - If `self` and `rhs` have unmatching bit widths.
-    pub fn into_wrapping_sub(self, rhs: &ApInt) -> Result<ApInt> {
-        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_sub_assign)
+    pub fn overflowing_uadd_assign(&mut self, rhs: &ApInt) -> Result<bool> {
```

Comment:
Most of the code there won't be useful anywhere else except if I implement `overflowing_uadd_assign`. I might however do something about the loops in this file. You see, I have barely used any iterators in my code because of the massive amount of special cases, preconditions, postconditions, and other things. There is a very common thing I use that involves

```
                        //pre loop carry setups with some function
                        let (temp, mut carry) = lhs[0].carrying_add(rhs[0]);
                        lhs[0] = temp;
                        for i in 1..(lhs.len() - 1) {
                            //main for body using the carries and the same function
                            let temp = lhs[i].dd().wrapping_add(rhs[i].dd()).wrapping_add(carry.dd());
                            lhs[i] = temp.lo();
                            carry = temp.hi();
                        }
                        //post loop no carry set part
                        let temp = lhs[lhs.len() - 1]
                            .wrapping_add(rhs[lhs.len() - 1])
                            .wrapping_add(carry);
```
And I will probably make some kind of macro or method for `Digit`s for that. Another reason for not using iterators is because I don't entirely trust that the compiler can optimize away certain things, and it is harder to reason about iterators when there is a lot of lazy evaluation, closures, borrowing, and dereferencing flying around.

Reply:
I have made many performance tests about iterators because I also didn't believe that the Rust compiler is able to optimize them as well as other constructs but in the end iterator and especially high-level programming in general was nearly always the way that the compiler optimized best in release mode. Of course in debug mode it oftentimes was the slowest but we do not mind for debug mode and care only for release mode when looking at performance.

Reply:
The logic behind this is: The compiler has an easier life determining the logic behind high-level code and a much harder time doing this for low-level code constructs.